### PR TITLE
GH-1229 Form Input Refactor

### PR DESCRIPTION
**Pull request by mlzummo**
Fixes-1229

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

**Comment by mlzummo** [comment]
@GordonSmith @jbrundage @dtsnell4 - please review

**Comment by dtsnell4** [comment]
Looks OK to me

File: src/form/Button.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            });
+        });
+
+        this._inputElement.forEach(function(e, idx) {
+            e.on("change", function (w) {
+                context.value([e.property("value")]);
+                w.change(w);
+            });
+        });
+    };
```
Comment (GordonSmith) [inline src/form/Button.js]: Could be simplified to:  
```
        this._inputElement[0] = element.append("button")
            .on("click", function (w) {
                w.click(w);
            })
            .on("blur", function (w) {
                w.blur(w);
            })
            on("change", function (w) {
                context.value([e.property("value")]);
                w.change(w);
            })
        ;
```

File: src/form/Button.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        var context = this;
+        this._inputElement.forEach(function(e) {
+            e.attr("name", context.name());
+        });
+
+        this.checkNodeName("BUTTON", element);
+        this._inputElement[0].text(this.value());
+    };
+
+    Button.prototype.createInput = function (element) {
```
Comment (GordonSmith) [inline src/form/Button.js]: should be in enter?  Not sure you really need a diff method here?

File: src/form/Button.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        this._inputElement.forEach(function(e, idx) {
+            e.on("change", function (w) {
+                context.value([e.property("value")]);
+                w.change(w);
+            });
+        });
+    };
+
+    Button.prototype.checkNodeName = function (expected, element) {
```
Comment (GordonSmith) [inline src/form/Button.js]: could be commoned up into a standard base class for all inputs?

**Comment by GordonSmith** [comment]
MIssing tests for some of the new Inputs (like Button)

File: src/form/Button.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        if (remove) {
+            this._inputElement.forEach(function(e) {
+               e.node().remove();
+            });
+            this.createInput(element);
+        }
+    };
+
+    Button.prototype.resetValue = function (w) {
+        w.value(w._inputElement.node().value);
```
Comment (GordonSmith) [inline src/form/Button.js]: Is _inputElement not an array now?

**Comment by GordonSmith** [comment]
(I only reviewed the first part) - looks like you have a lot of cut and paste code:
* disable
* checkNodeName 
* resetValue 
would be better if it could be moved to a common base class.

Also see my remarks about the creatInput method and how it could be simplified for Button - probably applicable to all the widgets?

File: src/form/Button.js (hunk 2 above)
Comment (mlzummo) [inline src/form/Button.js]: @GordonSmith are you talking about the createInput method?

Comment (GordonSmith) [inline src/form/Button.js]: Yes - why do we have a "createInput" method?  The code it contains is typically the code we have in the "enter" method, so would be better there?

**Comment by mlzummo** [comment]
@GordonSmith @jbrundage  - please re-review

**Comment by jbrundage** [comment]
This looks good to me

File: src/form/Select.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        this._inputElement[0] = element.append("select")
+            .on("click", function (w) {
+                w.click(w);
+            })
+            .on("blur", function (w) {
+                w.blur(w);
+            })
+            .on("change", function (w) {
+                context.value([this._inputElement[0].property("value")]);
```
Comment (jbrundage) [inline src/form/Select.js]: @mlzummo Change **this** to **context**

File: src/form/Button.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        
+        this._inputElement[0] = element.append("button")
+            .on("click", function (w) {
+                w.click(w);
+            })
+            .on("blur", function (w) {
+                w.blur(w);
+            })
+            .on("change", function (w) {
+                context.value([this._inputElement[0].property("value")]);
```
Comment (jbrundage) [inline src/form/Button.js]: @mlzummo Change **this** to **context**

**Comment by mlzummo** [comment]
@GordonSmith bump

File: src/api/IInput.js [hunk 7]
```diff
@@ -56,5 +56,29 @@
...
+            if (e.node().nodeName !== expected) {
+                return true;
+            }
+        });
+        if (remove) {
+            this._inputElement.forEach(function(e) {
+               e.node().remove();
+            });
+            this.createInput(element);
+        }
```
Comment (GordonSmith) [inline src/api/IInput.js]: Why do we need the whole "checkNodeName"?  Do we expect input types to change after they have been created?

File: src/api/IInput.js [hunk 8]
```diff
@@ -56,5 +56,29 @@
...
+        var remove = this._inputElement.some(function(e) {
+            if (e.node().nodeName !== expected) {
+                return true;
+            }
+        });
+        if (remove) {
+            this._inputElement.forEach(function(e) {
+               e.node().remove();
+            });
+            this.createInput(element);
```
Comment (GordonSmith) [inline src/api/IInput.js]: This call will fail for Button

File: src/api/IInput.js (hunk 7 above)
Comment (mlzummo) [inline src/api/IInput.js]: i think that was the original concept yes ... if we dont need it we can remove... thoughts?

File: src/api/IInput.js (hunk 8 above)
Comment (mlzummo) [inline src/api/IInput.js]: ill take a look at that

File: src/api/IInput.js (hunk 7 above)
Comment (GordonSmith) [inline src/api/IInput.js]: @jbrundage - thoughts (I think you added this check originally?)

Comment (jbrundage) [inline src/api/IInput.js]: @GordonSmith Originally, yes, the Input widget was built with the intention of an "input" being able to change "type" (ex: from a "text" input to a textarea).

But I don't think this concept works anymore with the latest implementation of form inputs (I don't think its necessary).

**Comment by mlzummo** [comment]
before pulling need to look at ․Invalid value for 'type':  checkbox expected set

**Comment by mlzummo** [comment]
@GordonSmith i think this is ready to pull now

**Comment by mlzummo** [comment]
do i need to make changes to marshaller?

**Pull request merged**